File: frontend/src/types/nomad-metrics.ts
export type Carrier = "GP" | "Robi" | "Banglalink" | "Teletalk";
export type SignalStrength = "No Signal" | "2G" | "3G" | "4G" | "5G";

export interface CarrierSignal {
  carrier: string;
  signal: string;
  votes: number;
}

export interface NomadMetricSummary {
  entry_id: string;
  avg_safety_rating: number | null;
  bkash_available_pct: number;
  signal_by_carrier: CarrierSignal[];
  has_submitted: boolean;
}

export interface NomadMetricSubmit {
  entry_id: string;
  carrier: Carrier;
  signal_strength: SignalStrength;
  safety_rating: number;
  bkash_available: boolean;
}

File: backend/app/schemas/nomad_metrics.py
import uuid
from pydantic import BaseModel, Field

class NomadMetricSubmit(BaseModel):
    entry_id: uuid.UUID
    carrier: str = Field(..., pattern="^(GP|Robi|Banglalink|Teletalk)$")
    signal_strength: str = Field(..., pattern="^(No Signal|2G|3G|4G|5G)$")
    safety_rating: int = Field(..., ge=1, le=5)
    bkash_available: bool

class CarrierSignal(BaseModel):
    carrier: str
    signal: str
    votes: int

class NomadMetricSummary(BaseModel):
    entry_id: uuid.UUID
    avg_safety_rating: float | None
    bkash_available_pct: int
    signal_by_carrier: list[CarrierSignal]
    has_submitted: bool = False